### ISSUE TITLE
Drop Node 4 support (and Node 6 and Node 8)

### ISSUE THREAD

**Issue by lipis** [prettier/prettier#3919]
The official support for Node 4 is ending soon ([April 2018](https://github.com/nodejs/Release#nodejs-release-working-group)) and from time to time we have quite a few issues like #3875 with the failed automated tests.

What can possibly go wrong?! 

(It was also discussed in the #3503, but I'm not sure how far we are with that one)

**Comment by lipis** [comment]
Already discussed: https://github.com/prettier/prettier/issues/3503#issuecomment-359937346

**Comment by j-f1** [comment]
I still think we should do this, although we might want to give people a few months to migrate away from the soon-to-be-insecure Node 4. On the other hand, since Prettier doesn’t go into production code (unless I’m missing something), we could move on and say that people who need Node 4 support can just use an older version of Prettier with Node 4 support until they can migrate.

**Comment by evilebottnawi** [comment]
I agree, some parsers can require minimum nodejs `6` and problems will be corrected only for the version `6` version

**Comment by duailibe** [comment]
I think we can "oficially" drop support for node 4, but not necessarily we have to make it stop working for node 4.

**Comment by j-f1** [comment]
How about switching Travis to 6/stable instead of 4/stable?

**Comment by TrySound** [comment]
`stable` is deprecated use `node` instead
https://github.com/creationix/nvm#usage

**Comment by j-f1** [comment]
Thanks for pointing that out @TrySound! Fixed.

**Comment by duailibe** [comment]
@j-f1 Actually I was thinking about still running node 4 tests, but maybe installing the dependencies with `--ignore-engines` (only for node 4) and see if it still works.

**Comment by azz** [comment]
A concern is that libs using Prettier still support Node 4, and they'll be running Prettier in their CI on Node 4.

**Comment by TrySound** [comment]
Maybe this users should just use current major instead of the next one?

**Comment by suchipi** [comment]
I don't think it's safe to drop node 4 support without bumping to 2.0.

**Comment by suchipi** [comment]
That sounds reasonable. Then we could transpile for npm, but only have the src in the git repo. Great solution

**Comment by brodybits** [comment]
I am looking forward to this idea, which would need to be done in a major release such as 2.0 (#3503). I find it really ugly that `package.json` specifies `"node": ">=6"` while generated `dist/package.json` specifies `"node": ">=4"`.

**Comment by duailibe** [comment]
@brodybits why is it ugly? I believe our current solution works fine

**Comment by brodybits** [comment]
Let me rephrase it to "cryptic (non-obvious)" then (I already updated my comment). I am looking forward to the new release, with no Node.js 4 support baggage. A side point is that Node.js 6 maintenance will end pretty soon, as I reported in #5711.

**Comment by j-f1** [comment]
We’ll probably drop Node 6 support when Node does, but not for the production version.

**Comment by duailibe** [comment]
I believe the majority of our users won't even know (or care) that we have different configurations. I'm obviously biased but I don't see any downsides in doing that (or benefits removing it).

I don't have any data on Node 4 usage, but if it's not too much trouble to continue supporting it for the near future, I don't see a problem with it

**Comment by rattrayalex** [comment]
Coming from @jessidhia 's [comment](https://github.com/prettier/prettier/issues/3503#issuecomment-459638439) on another thread, I think it'd be appropriate to hard-break Node 4 in prettier 2.0 (ie; with an explicit error) and even to consider doing the same for Node 6, [which hits EOL this April](https://github.com/nodejs/Release).

Prettier community members are unlikely to be _forced_ to upgrade to prettier 2.0, so I don't think we'll be causing too much pain by doing this. On the other hand, running an insecure version of Node is bad for both users and the JS ecosystem; if we can use a new version of prettier as a carrot/incentive for people to upgrade, I think we should.

**Comment by brodybits** [comment]
From <https://github.com/prettier/prettier/issues/5711#issuecomment-519258262> I gather that Node.js 6 will be dropped as well.

I would also favor dropping Node.js 8 with [EOL coming up at the end of December](https://github.com/nodejs/Release).

I hope this will be done soon. Supporting old Node.js versions makes it harder to keep the dependencies up to date (see PR #6224 for example).

**Comment by fisker** [comment]
I think we can still make production support old versions, since we use rollup and webpack bundle eveything, but drop the source support for old nodejs versions, we can even switch source to esm

**Comment by fisker** [comment]
I really wonder is there anyone install prettier from this repo instead of install from npm? we should not support that

**Comment by duailibe** [comment]
Dropping Node 4 is already planned, but it's a breaking change, it'll go out with 2.0. It needs to be coordinated.

> Supporting old Node.js versions makes it harder to keep the dependencies up to date

@brodybits What are the dependencies that are incompatible and why do we need to update them?

**Comment by duailibe** [comment]
> we should not support that

@fisker Why not?

**Comment by brodybits** [comment]
> @brodybits What are the dependencies that are incompatible and why do we need to update them?

get-stream@5.1.0 was one; I thought there were some others. I would have to check again.

Outdated dependencies can miss security updates and other bug fixes. Outdated dependencies can also be harder to update in the future.

>> there anyone install prettier from this repo
>> [...]
> > we should not support that
> 
> @fisker Why not?

- additional testing burden
- makes it harder to update dependencies without breaking this functionality

I would personally favor we support this functionality for recent Node.js versions only.

**Comment by fisker** [comment]
> Outdated dependencies can also be harder to update in the future

Agree, last time when I try to upgrade `rollup`, I have to look very old docs issues and changlog

**Comment by fisker** [comment]
> we'd like to easy support for forking Prettier and so people can just fork and install directly from their fork

why not fork, change, build, install from bundle? this is what I would do, if I want to change something

**Comment by duailibe** [comment]
> Outdated dependencies can miss security updates and other bug fixes. Outdated dependencies can also be harder to update in the future.

@brodybits Maybe, but practically that hasn't happened. All the "security updates" GitHub warned so far didn't affect us at all. Specially since this project is just a formatter.

> I would personally favor we support this functionality for recent Node.js versions only.

@brodybits There may be people running a forked Prettier in CI's with old Node.js versions.

> Agree, last time when I try to upgrade rollup, I have to look for very old docs and issues and changlog

@fisker Was it necessary to update rollup though? Did we have any benefits from upgrading it?

> why not fork, change, build, install from bundle?

@fisker How to install from bundle? Where will you keep this bundle?

**Comment by fisker** [comment]
> Was it necessary to update rollup though? Did we have any benefits from upgrading it?

You're the one who want it updated https://github.com/prettier/prettier/issues/6154 https://github.com/prettier/prettier/pull/6200

**Comment by evilebottnawi** [comment]
@duailibe some parsers can drop support node@4 and it will be problem for us

**Comment by duailibe** [comment]
@fisker Forgot about that 👍 The comment is valid for everything else

@evilebottnawi There are some dependencies that have dropped support and that doesn't affect us because we transpile everything.

And I'll repeat:

> Dropping Node 4 is already planned, but it's a breaking change, it'll go out with 2.0. It needs to be coordinated.

**Comment by j-f1** [comment]
@duailibe You dropped Node 4 support for installing Prettier directly from GitHub. Do you think it would be reasonable to also drop Node 6 support for direct-from-GitHub installs?

**Comment by brodybits** [comment]
> Do you think it would be reasonable to also drop Node 6 support for direct-from-GitHub installs?

I would *personally* consider this to be a breaking change if we want to be strictly semver compliant, which I would favor.

I really hope we can start targeting the new major release and update the dependencies soon. Supporting the old Node.js versions seems to divert attention and resources away from highly desired updates and bug fixes.

**Comment by evilebottnawi** [comment]
@brodybits we already doesn't support node@4 https://github.com/prettier/prettier/blob/master/package.json#L14

/cc @prettier/core I think we can close issue

**Comment by thorn0** [comment]
> we already doesn't support node@4

@evilebottnawi [we do](https://github.com/prettier/prettier/blob/b5cc13a5dfa92e9166a47b9daf8cce74daeb7722/scripts/build/build.js#L80), but not for direct installation from GitHub

**Comment by evilebottnawi** [comment]
@thorn0 Some of our deps already require node@6, so i think we can't work on node@4

**Comment by thorn0** [comment]
@evilebottnawi The tests pass 🤷‍♂️

**Comment by evilebottnawi** [comment]
hm :confused:

**Comment by thorn0** [comment]
Exactly. It's the version that is released to [npm](https://unpkg.com/browse/prettier@1.18.2/package.json). It works in Node 4 because it and all its dependencies are transpiled.

**Comment by evilebottnawi** [comment]
@thorn0 :+1: I think we can close this issue

**Comment by thorn0** [comment]
@evilebottnawi Node 4 is still supported in 1.x. This issue is part of the 2.0 milestone. We can close it when we have a branch for 2.0 and a commit there that changes [this](https://github.com/prettier/prettier/blob/b5cc13a5dfa92e9166a47b9daf8cce74daeb7722/scripts/build/build.js#L80).

**Comment by evilebottnawi** [comment]
@thorn0 yep, you are right

**Comment by sosukesuzuki** [comment]
#6907 #6908 #7302 has been merged into `next`.
What else should we do for this issue? Or can we close this issue?
